### PR TITLE
optimize groupStack performance

### PR DESCRIPTION
**Pull request by ChiwanAhn**
### **User description**
**Issue:**
The timeline library experienced performance bottlenecks when handling a large number of items, causing the app to freeze. The core issue was the inefficient placement of overlapping items, leading to excessive calculations and improper stacking.

**Solution:**
This PR introduces an optimized item placement logic to address the performance issues:

1. Efficient Overlap Detection
  - Implemented an interval tree to quickly and efficiently detect overlapping items.
  - This reduces the complexity of finding overlapping items, which is crucial for large datasets.

2. Optimized Placement Logic:
  - Sorted colliding items by their top position
  - Identified and utilized the first available vertical space to place new items, ensuring proper stacking and reducing unnecessary calculations.

3. Improved Gap Identification:
  - Enhanced the logic to find gaps between colliding items.
  - Placed items in the first available gap, only resorting to stacking below if no suitable gap was found.

performance benchmark between current version(v1) and new version(v2) on unit test
Screenshot 2024-05-31 at 2 37 32 AM

performance test on demo/performance page
- tested with 30k items
- moved prev, next year to trigger stack recalculation

Before
took +600ms 
image

After
took +80ms
image

https://github.com/LifeCoded/react-calendar-timeline/assets/25087511/d594aa45-3067-4650-b049-0a785e9081e5

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Introduced `IntervalTree` class for efficient interval management and optimized item stacking logic.
- Added comprehensive unit tests for `groupStack` function and `IntervalTree` class.
- Enhanced performance demo with realistic data generation and updated time handling logic.
- Fixed hot module replacement rendering issue in the demo.
- Updated package configuration, including version bump to `1.0.0` and TypeScript support.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Tests
</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>group-stack.js</strong><dd><code>Add performance and unit tests for `groupStack` function.</code></dd></summary>
<hr>
      
__tests__/utils/calendar/group-stack.js

<li>Added performance tests for <code>groupStack</code> function.<br> <li> Introduced new test cases for stacking logic.<br> <li> Implemented a helper function to generate dimension items.<br>

</details>
    

  </td>
  <td><a href="https://github.com/LifeCoded/react-calendar-timeline/pull/4/files#diff-4c5faca8f5382b9f85346a907069186f1009d89f34d7178fe8ba5aa78f8538a2">+130/-10</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>intervalTree.test.js</strong><dd><code>Add unit tests for `IntervalTree` class.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
__tests__/utils/intervalTree/intervalTree.test.js

<li>Added tests for <code>IntervalTree</code> class.<br> <li> Verified interval insertion and overlap detection.<br>

</details>
    

  </td>
  <td><a href="https://github.com/LifeCoded/react-calendar-timeline/pull/4/files#diff-921593fcc1ea0ed9eaec162d2b15ab9e412d337cb9ed33813d8c656d75717f24">+65/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Configuration changes
</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>jest-setup.js</strong><dd><code>Configure Jest to include performance hooks.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
config/jest-setup.js

<li>Imported <code>performance</code> from <code>perf_hooks</code>.<br> <li> Added <code>performance</code> to global scope.<br>

</details>
    

  </td>
  <td><a href="https://github.com/LifeCoded/react-calendar-timeline/pull/4/files#diff-0d3b990aefab39f0b0154c181747772d706c181e70eb3c4f18b327450ae67b09">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>package.json</strong><dd><code>Update package configuration and version.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
package.json

<li>Updated version to <code>1.0.0</code>.<br> <li> Added <code>types</code> field for TypeScript support.<br> <li> Ignored <code>node_modules</code> and <code>test-utility</code> in Jest watch paths.<br>

</details>
    

  </td>
  <td><a href="https://github.com/LifeCoded/react-calendar-timeline/pull/4/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519">+7/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Enhancement
</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>index.js</strong><dd><code>Update performance demo with realistic data generation.</code>&nbsp; &nbsp; </dd></summary>
<hr>
      
demo/app/demo-performance/index.js

<li>Updated demo to generate more realistic fake data.<br> <li> Modified time handling logic for performance testing.<br>

</details>
    

  </td>
  <td><a href="https://github.com/LifeCoded/react-calendar-timeline/pull/4/files#diff-4416807b116680b38d4f6084d90f6bac6029f847e0cbf7ce66d0f599f72f5815">+13/-21</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>generate-fake-data.js</strong><dd><code>Enhance fake data generation for performance testing.</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
demo/app/generate-fake-data.js

<li>Adjusted fake data generation logic for better performance testing.<br>

</details>
    

  </td>
  <td><a href="https://github.com/LifeCoded/react-calendar-timeline/pull/4/files#diff-300911f3aa224393bd5c5581047f21d266789487ddd35a95e456d1f553816aba">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>calendar.js</strong><dd><code>Optimize item stacking logic using interval tree.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
src/lib/utility/calendar.js

<li>Replaced collision detection logic with interval tree.<br> <li> Optimized <code>groupStack</code> function for better performance.<br> <li> Updated <code>stackGroup</code> to use interval tree.<br>

</details>
    

  </td>
  <td><a href="https://github.com/LifeCoded/react-calendar-timeline/pull/4/files#diff-5dbb0fe5624ad16cd67ea02ab4c50b3992f8f389e0ec71e8ab10b383957c21d5">+71/-72</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>intervalTree.js</strong><dd><code>Implement `IntervalTree` class for interval management.</code>&nbsp; &nbsp; </dd></summary>
<hr>
      
src/lib/utility/intervalTree.js

<li>Implemented <code>IntervalTree</code> class for efficient interval management.<br> <li> Added methods for interval insertion and overlap searching.<br>

</details>
    

  </td>
  <td><a href="https://github.com/LifeCoded/react-calendar-timeline/pull/4/files#diff-b892fe922a5c5b03e4daad417b7d4c8032e74b512a439a66cebfdc43a3cc4423">+188/-0</a>&nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Bug fix
</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>index.js</strong><dd><code>Fix hot module replacement rendering in demo.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
demo/index.js

- Fixed hot module replacement rendering issue.

</details>
    

  </td>
  <td><a href="https://github.com/LifeCoded/react-calendar-timeline/pull/4/files#diff-94c7feea3c21222ef6b9117a4e177efc289063182228557520c91c1e06fa238f">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

----

| :rocket: | This description was created by [Ellipsis](https://www.ellipsis.dev) for commit 7c5663a967f0eafccac3a80741049f2398309955  | 
|--------|--------|

### Summary:
This PR optimizes the `groupStack` function in the react-calendar-timeline library by introducing an `IntervalTree` class, enhancing performance, and updating demos and configurations.

**Key points**:
- Introduced `IntervalTree` class for efficient interval management and optimized item stacking logic.
- Optimized `groupStack` function for better performance.
- Added comprehensive unit tests for `groupStack` function and `IntervalTree` class.
- Enhanced performance demo with realistic data generation and updated time handling logic.
- Updated package configuration, including version bump to `1.0.0` and TypeScript support.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

**Comment by ChiwanAhn** [comment]
bumped version to 1.0.0 for easier version maintenance.

File: package.json [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
   "name": "@mosaicapp/react-calendar-timeline",
-  "version": "0.27.3-react18",
+  "version": "1.0.0",
   "description": "react calendar timeline",
   "publishConfig": {
     "registry": "https://registry.npmjs.org/"
   },
   "main": "lib/index.js",
-    "scripts": {
+  "types": "lib/index.d.ts",
```
Comment (ChiwanAhn) [inline package.json:9]: it is problematic changes on the web repo side.

revert this change and bump the version again to 1.0.1 (should've bumped to v1 after checks but caught while checking types on web repo)

File: src/lib/utility/calendar.js [hunk 2]
```diff
@@ -256,72 +257,69 @@ export function getVisibleItems(items, canvasTimeStart, canvasTimeEnd, keys) {
...
-  group,
-  groupHeight,
-  itemIndex
-) {
-  // calculate non-overlapping positions
-  let curHeight = groupHeight
-  let verticalMargin = (lineHeight - item.dimensions.height) / 2
+export function groupStack(lineHeight, item, intervalTree, groupHeight, itemIndex) {
+  let curHeight = groupHeight;
+  const verticalMargin = (lineHeight - item.dimensions.height) / 2
```
Comment (profaurore) [inline src/lib/utility/calendar.js:272]: Can you add a comment here: `Vertical space above and below each item.`

**Review by profaurore — APPROVED** [review]
Just some minor comments.

File: src/lib/utility/calendar.js [hunk 3]
```diff
@@ -256,72 +257,69 @@ export function getVisibleItems(items, canvasTimeStart, canvasTimeEnd, keys) {
...
+    if (collidingItems.length > 0) {
+      // Sort colliding items by their top position.
+      collidingItems.sort((a, b) => a.data.top - b.data.top);
+
+      // Try to place the item in the first available gap.
+      let placed = false;
+      let previousBottom = 0;
+
+      for (let i = 0; i < collidingItems.length; i++) {
+        let collidingItem = collidingItems[i];
```
[the remark this answers is not in the document]
Comment (ChiwanAhn) [reply src/lib/utility/calendar.js]: addressed in [ec4ee9a](https://github.com/LifeCoded/react-calendar-timeline/pull/4/commits/ec4ee9a0a5918e5dda3af0f9108442f4ad4aafb7)

not going to publish with this change as it is minor refactoring that doesn't change any behaviour

File: src/lib/utility/calendar.js (hunk 2 above)
Comment (ChiwanAhn) [reply src/lib/utility/calendar.js:273]: resolved in [ec4ee9a](https://github.com/LifeCoded/react-calendar-timeline/pull/4/commits/ec4ee9a0a5918e5dda3af0f9108442f4ad4aafb7)

**Pull request merged**